Add custom_aspect parameter to plot_point_brain.

VinceBaz opened the pull request:
The function plot_point_brain() sets the limits of the xyz axes with the following lines of code:

`        ax.set(xlim=0.57 * np.array(ax.get_xlim()),
               ylim=0.57 * np.array(ax.get_ylim()),
               zlim=0.60 * np.array(ax.get_zlim()))`

It also zscores the coordinates beforehand.

This might be problematic since these limits, being hard-coded, can lead to various bugs. For instance, if we use the coordinates of the Lausanne 1000 parcellation, some parcels are cropped out:

![image](https://user-images.githubusercontent.com/46877241/122461313-2bf2df00-cf81-11eb-8673-e6ef42c2a619.png)

To solve this particular bug, I modified those lines to 

`                ax.set(xlim=0.665 * np.array(ax.get_xlim()),
                       ylim=0.665 * np.array(ax.get_ylim()),
                       zlim=0.70 * np.array(ax.get_zlim()))`.

We now get this when we try to plot the Lausanne 1000 parcellation:

![image](https://user-images.githubusercontent.com/46877241/122461541-783e1f00-cf81-11eb-964a-75f940b1c635.png).

The parcels are no longer cropped out.

 Also, I personally enjoy visualizing my data such that the axes are automatically scaled to have "equal" aspect ratios, so I added a custom_aspect parameter to the function that can be set to False if the user wants. If this parameter is set to False, then the axes will be automatically scaled, which gives us:

![image](https://user-images.githubusercontent.com/46877241/122461872-e5ea4b00-cf81-11eb-8a39-3d70bd79b9ce.png).

This comes with the caveat that there is more space between the differerent views, and now the 3 brains are not perfectly aligned on top of one another. Also, the sagittal view is now much more elongated, which might be dissapointing if the user wants to plot a "toy" version of the brain that looks nice. So for this reason, this custom_aspect parameter is set to True by default (which gives us what we had before).

liuzhenqi77 commented:
Hi Vince, I think since last year, `plot_point_brain` were not displaying the right perspective/shape, and these fixed numbers (to set the limit) might be for the previously-working version. Ross @rmarkello has made some changes #68 to remove the aspect parameter, but at the time it still wouldn't work nicely because there was a bug in matplotlib/Axes3D (see matplotlib/matplotlib#1077, matplotlib/matplotlib#8894, matplotlib/matplotlib#17172), and there were quite some effects to fix it (see matplotlib/matplotlib#8896, matplotlib/matplotlib#16472, matplotlib/matplotlib#17515). 

Now it seems 3d aspect is fixed in [3.3.0](https://matplotlib.org/stable/users/prev_whats_new/whats_new_3.3.0.html#d-axes-improvements) using [`set_box_aspect`](https://matplotlib.org/stable/api/_as_gen/mpl_toolkits.mplot3d.axes3d.Axes3D.html#mpl_toolkits.mplot3d.axes3d.Axes3D.set_box_aspect) from matplotlib/matplotlib#17515, we could maybe work a bit to fully fix the point brain that we all need!

It also seems `aspect="equal"` (matplotlib/matplotlib#1077) is [still not fixed](https://github.com/matplotlib/matplotlib/issues/1077#issuecomment-853567145) until now.

VinceBaz commented:
Ohh Thanks @liuzhenqi77. I did remember that Ross modified the code because at some point, they decided to remove the implementation of `ax.set_aspect('equal')` altogether: https://github.com/matplotlib/matplotlib/pull/13474. It turns out that replacing the line 

`ax.auto_scale_xyz(*[[np.min(scaling), np.max(scaling)]] * 3)`

to 
`ax.set_box_aspect(tuple(scaling[:, 1] - scaling[:, 0]))`

makes our brains look better: we scale the box aspect (as the name of the function suggests) rather than directly set the limits of the axes. So basically, we get from this (the modified version I proposed above; I added the axes so that we can see what is happening):

![image](https://user-images.githubusercontent.com/46877241/122569225-8723e080-d018-11eb-8899-99e2be047f11.png)

to this:

![image](https://user-images.githubusercontent.com/46877241/122569370-aae72680-d018-11eb-901a-18340ade01db.png)

I think that scaling the box aspect is better. And for some reason, the sagittal view now looks much better.

VinceBaz commented:
To confirm that the set_box_aspect() function works as expected, I looked at what the aspect would look like, for the sagittal view, in 2d space, with aspect='equal':

![image](https://user-images.githubusercontent.com/46877241/122573009-5180f680-d01c-11eb-80fd-6dc8e1847118.png).

It looks exactly the same. Since the three views are now much better, I think that we should always want to set the aspect ratio to equal. So I updated by pull request and got rid of the "custom_aspect" parameter and as I mentionned in my previous comment, I added a views_orientation parameter as well.

rmarkello commented:
This is awesome! :tada: :raised_hands: Thanks @VinceBaz for the PR and @liuzhenqi77 for digging through the myriad matplotlib issues + errors + updates. 

I'm happy to take a look at the code but since @liuzhenqi77 was here first maybe you want to take lead on the review? :grin: I've gone ahead and assigned it to you but let me know if you'd prefer not to.

liuzhenqi77 commented on netneurotools/plotting.py:
```diff
@@ -761,18 +764,22 @@ def plot_point_brain(data, coords, views=None, cbar=False, figsize=(4, 4.8),
+    if views_orientation == 'vertical':
+        ncols, nrows = 1, len(views)
+    elif views_orientation == 'horizontal':
+        ncols, nrows = len(views), 1
```
Do you think it would make sense to also change (switch x/y) the figsize if not set explicitly by the user?

VinceBaz replied:
Hum, I am wondering if its ok to have the default parameter being changed without the user being explicitly told about the change. I.E. currently, we set the default figsize to (4, 4.8). If we then change it to (4.8, 4), it might be confusing for the user. What I think could be a good idea is to set this default parameter to None, and then automatically adjust the figsize depending on a) the number of views the user wants, and the orientation of the views. For example, if the user wants two views, and horizontal, the figsize would be set to (9.6, 4).

liuzhenqi77 replied:
Good idea, I agree!

VinceBaz replied:
I replaced the figsize parameter by the views_size parameter (so as not to confuse the user). The actual figsize parameter is adjusted according to the number of views and the orientation of the different views.

liuzhenqi77 reviewed (commented):
Tested locally and works nicely! Thank you Vince!
We may also add matplotlib>=3.3.0 to the requirements.txt

liuzhenqi77 commented:
Looks great to me! Pinging @rmarkello
